File: examples/governance_static_go/domain/retire_model.go
package domain

import (
	"time"
	"fmt"
)

type RetireModel struct {
	ModelId string `json:"model_id"`
}

func (c RetireModel) CommandName() string { return "RetireModel" }

func (c RetireModel) Execute(repo AiModelRepository) (*AiModel, *RetiredModel, error) {
	existing, err := repo.Find(c.ModelId)
	if err != nil {
		return nil, nil, err
	}
	if existing == nil {
		return nil, nil, fmt.Errorf("AiModel not found: %s", c.ModelId)
	}
	if existing.Status != "approved" && existing.Status != "suspended" {
		return nil, nil, fmt.Errorf("cannot RetireModel: AiModel is in %s state", existing.Status)
	}
	existing.Status = "retired"
	existing.UpdatedAt = time.Now()
	if err := existing.Validate(); err != nil {
		return nil, nil, err
	}
	if err := repo.Save(existing); err != nil {
		return nil, nil, err
	}
	event := RetiredModel{
		AggregateID: existing.ID,
		ModelId: c.ModelId,
		OccurredAt: time.Now(),
	}
	return existing, &event, nil
}

File: examples/governance/model_registry_domain/lib/model_registry_domain/ai_model/commands/retire_model.rb
module ModelRegistryDomain
  class AiModel
    module Commands
      class RetireModel
        include Hecks::Command
        emits "RetiredModel"

        attr_reader :model_id

        def initialize(model_id: nil)
          @model_id = model_id
        end

        def call
          existing = repository.find(model_id)
          if existing
            unless ["approved", "suspended"].include?(existing.status)
              raise ModelRegistryDomain::Error, "Cannot RetireModel: status must be one of approved, suspended, got '#{existing.status}'"
            end
            AiModel.new(
              id: existing.id,
              name: existing.name,
              version: existing.version,
              provider_id: existing.provider_id,
              description: existing.description,
              risk_level: existing.risk_level,
              registered_at: existing.registered_at,
              parent_model_id: existing.parent_model_id,
              derivation_type: existing.derivation_type,
              capabilities: existing.capabilities,
              intended_uses: existing.intended_uses,
              status: "retired"
            )
          else
            raise ModelRegistryDomain::Error, "AiModel not found: #{model_id}"
          end
        end
      end
    end
  end
end
